builtin/logical/transit/path_rotate.go

```diff
@@ -31,26 +31,38 @@ func (b *backend) pathRotateWrite(
 	name := d.Get("name").(string)
 
 	// Get the policy
-	lp, err := b.policies.getPolicy(req, name)
+	p, lockType, err := b.lm.GetPolicy(req.Storage, name)
 	if err != nil {
 		return nil, err
 	}
-
-	// Error if invalid policy
-	if lp == nil {
-		return logical.ErrorResponse("policy not found"), logical.ErrInvalidRequest
+	if p == nil {
+		return logical.ErrorResponse("key not found"), logical.ErrInvalidRequest
 	}
 
-	lp.Lock()
-	defer lp.Unlock()
+	// Store so we can detect later if this has changed out from under us
+	keyVersion := p.LatestVersion
+
+	b.lm.UnlockPolicy(name, lockType)
+
+	// Refresh in case it's changed since before we grabbed the lock
+	p, err = b.lm.RefreshPolicy(req.Storage, name)
+	if err != nil {
+		return nil, err
+	}
+	if p == nil {
+		return nil, fmt.Errorf("error finding key %s after locking for changes", name)
+	}
+	defer b.lm.UnlockPolicy(name, exclusive)
 
-	// Verify if wasn't deleted before we grabbed the lock
-	if lp.policy == nil {
-		return nil, fmt.Errorf("no existing policy named %s could be found", name)
+	// Make sure that the policy hasn't been rotated simultaneously
+	if keyVersion != p.LatestVersion {
```

Comment:
I don't know much about programming in that style and don't have time to change everything around again right now.